tests/shouldwork/Cores/Xilinx/DcFifo/Abstract.hs

```diff
@@ -0,0 +1,151 @@
+  Clock dom -> Reset dom -> Enable dom ->
+  -- | Stall circuit? For this test case, this signal comes from 'lfsrF'
+  Signal dom Stall ->
+  -- | Signals from FIFO
+  Signal dom (Empty, DataCount depth, BitVector n) ->
+  -- | Maybe output read from FIFO
+  Signal dom (Bool, Maybe (BitVector n))
+fifoSampler clk rst ena stalls inps =
+  mealy clk rst ena go False (bundle (stalls, inps))
+ where
+  go ::
+    ReadLastCycle ->
+    (Stall, (Empty, DataCount depth, BitVector n)) ->
+    (ReadLastCycle, (Bool, Maybe (BitVector n)))
+  go readLastCycle (stall, (fifoEmpty, _dataCount, readData)) = (readNow, (readNow, maybeData))
+   where
+    maybeData = readLastCycle `orNothing` readData
+    readNow = not stall && not fifoEmpty
+{-# NOINLINE fifoSampler #-}
+
+-- | Drives Xilinx FIFO with an ascending sequence of 'BitVector's. Stalls
+-- intermittently based on stall input.
+fifoDriver ::
+  forall n dom depth .
+  ( KnownDomain dom
+  , KnownNat n ) =>
+  Clock dom -> Reset dom -> Enable dom ->
+  -- | Stall circuit? For this test case, this signal comes from 'lfsrF'
+  Signal dom Stall ->
+  -- | Signals from FIFO
+  Signal dom (Full, DataCount depth) ->
+  -- | Maybe write input to FIFO
+  Signal dom (Maybe (BitVector n))
+fifoDriver clk rst ena stalls inps =
+  mealyB clk rst ena go 0 (stalls, inps)
+ where
+  go ::
+    BitVector n ->
+    (Stall, (Full, DataCount depth)) ->
+    (BitVector n, Maybe (BitVector n))
+  go n0 (stall, (full, _dataCount)) = (n1, maybeWrite)
+   where
+    maybeWrite = willWrite `orNothing` n0
+    willWrite = not stall && not full
+    n1 = if willWrite then succ n0 else n0
+
+type ConfiguredFifo read write =
+  Clock write -> Clock read ->
+  -- | Asynchronous reset
+  Reset read ->
+
+  -- | Write data
+  Signal write (Maybe (BitVector 16)) ->
+  -- | Read enable
+  Signal read Bool ->
+  FifoOut read write 4 (BitVector 16)
+
+mkTestBench ::
+  forall read write.
+  (KnownDomain write, KnownDomain read) =>
+  ConfiguredFifo read write ->
+  Signal read Bool
+mkTestBench cFifo = done
+ where
+  (rClk, wClk) = biTbClockGen (not <$> done)
+
+  rRst = resetGen
+  wRst = resetGen
+
+  rEna = enableGen
+  wEna = enableGen
+
+  -- The reset is asynchronous; we make sure the reset is always asserted at
+  -- least one cycle in the slower clock domain.
+  slowReset =
+    if snatToNum @Natural (clockPeriod @write) > snatToNum @Natural (clockPeriod @read)
+      then unsafeToReset (unsafeSynchronizer wClk rClk (unsafeFromReset wRst))
+      else rRst
+
+  -- Driver
+  wLfsr = bitToBool <$> lfsrF wClk wRst wEna 0xDEAD
+  writeData = fifoDriver wClk wRst wEna wLfsr (bundle (isFull, writeCount))
+
+  -- Sampler
+  rLfsr = bitToBool <$> lfsrF rClk rRst rEna 0xBEEF
+  (readEnable, maybeReadData) =
+    unbundle $
+      fifoSampler rClk slowReset rEna rLfsr (bundle (isEmpty, readCount, fifoData))
+
+  FifoOut{isFull, writeCount, isEmpty, readCount, fifoData} =
+    cFifo wClk rClk rRst writeData readEnable
```

Comment:
This is the one you want to reset with `slowReset`, not the `fifoSampler`.

I think for the purposes of the test bench, it's okay to keep the rest of the circuit in reset for just the first cycle of the respective clock. In a synthesised circuit, I'd use more elaborate reset sequencing, but we're only interested in getting the Xilinx behavioural model behave right.